aiperf/records/records_manager.py

```diff
@@ -435,21 +463,88 @@ async def _on_profile_cancel_command(
-        """Report the real-time metrics at a regular interval (only if the UI type is dashboard)."""
+    @background_task(interval=DEFAULT_REALTIME_METRICS_INTERVAL, immediate=False)
+    async def _report_realtime_inference_metrics_task(self) -> None:
+        """Report inference metrics at regular intervals (dashboard only)."""
         if self.service_config.ui_type != AIPerfUIType.DASHBOARD:
             return
-        while not self.stop_requested:
-            await asyncio.sleep(DEFAULT_REALTIME_METRICS_INTERVAL)
-            async with self.processing_status_lock:
-                if (
-                    self.processing_stats.total_records
-                    == self._previous_realtime_records
-                ):
-                    continue  # No new records have been processed, so no need to update the metrics
-                self._previous_realtime_records = self.processing_stats.processed
-            await self._report_realtime_metrics()
+
+        async with self.processing_status_lock:
+            if self.processing_stats.total_records == self._previous_realtime_records:
+                return
+            self._previous_realtime_records = self.processing_stats.processed
+
+        metrics = await self._generate_realtime_metrics()
+        if metrics:
+            await self.publish(
+                RealtimeMetricsMessage(
+                    service_id=self.service_id,
+                    metrics=metrics,
+                )
+            )
+
+    @background_task(interval=DEFAULT_REALTIME_METRICS_INTERVAL, immediate=False)
+    async def _report_realtime_telemetry_metrics_task(self) -> None:
+        """Report telemetry metrics at regular intervals (dashboard + telemetry mode only)."""
+        if self.service_config.ui_type != AIPerfUIType.DASHBOARD:
+            return
+        if self.user_config.gpu_telemetry_mode != GPUTelemetryMode.REALTIME_DASHBOARD:
+            return
+
+        telemetry_metrics = await self._generate_realtime_telemetry_metrics()
+
+        self._telemetry_total_metrics_generated += len(telemetry_metrics)
+
+        if telemetry_metrics:
+            await self.publish(
+                RealtimeTelemetryMetricsMessage(
+                    service_id=self.service_id,
+                    metrics=telemetry_metrics,
+                )
+            )
+        else:
+            # Get telemetry hierarchy from processor
+            if not self._telemetry_results_processors:
+                return
+
+            processor = self._telemetry_results_processors[0]
+            telemetry_hierarchy = processor.get_telemetry_hierarchy()
+
+            endpoints_with_data = set(telemetry_hierarchy.dcgm_endpoints.keys())
+
+            self.info(
+                f"Configured endpoints: {self._configured_telemetry_endpoints}, "
+                f"Endpoints with data: {endpoints_with_data}"
+            )
+
+            for endpoint in self._configured_telemetry_endpoints:
```

Comment:
this whole for loop + if-else blocks is a bit unruly and could use some cleaning up, i think